// node-opcua-address-space/dist/source/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeIdManager = exports.instantiateCertificateExpirationAlarm = exports.ensureDatatypeExtractedWithCallback = exports.ensureDatatypeExtracted = exports.ContinuationPointManager = exports.resolveOpaqueOnAddressSpace = exports.adjustBrowseDirection = exports.promoteToTwoStateDiscrete = exports.promoteToMultiStateValueDiscrete = exports.promoteToMultiStateDiscrete = void 0;
/**
 * @module node-opcua-address-space
 */
__exportStar(require("./address_space_ts"), exports);
__exportStar(require("./interfaces/state_machine/ua_state_machine_type"), exports);
__exportStar(require("./interfaces/state_machine/ua_exclusive_limit_state_machine_type_ex"), exports);
__exportStar(require("./interfaces/state_machine/ua_finite_state_machine_type"), exports);
__exportStar(require("./interfaces/state_machine/ua_program_state_machine_type"), exports);
__exportStar(require("./interfaces/state_machine/ua_shelved_state_machine_ex"), exports);
__exportStar(require("./interfaces/state_machine/ua_transition_ex"), exports);
__exportStar(require("./interfaces/ua_subscription_diagnostics_variable_ex"), exports);
__exportStar(require("./interfaces/data_access/ua_multistate_discrete_ex"), exports);
__exportStar(require("./interfaces/data_access/ua_multistate_value_discrete_ex"), exports);
__exportStar(require("./interfaces/data_access/ua_two_state_discrete_ex"), exports);
__exportStar(require("./interfaces/data_access/ua_y_array_item_ex"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/condition_info_i"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/condition_snapshot"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/instantiate_alarm_condition_options"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/instantiate_condition_options"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/instantiate_exclusive_limit_alarm_options"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/instantiate_limit_alarm_options"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/instantiate_off_normal_alarm_options"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/ua_acknowledgeable_condition_ex"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/ua_alarm_condition_ex"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/ua_certificate_expiration_alarm_ex"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/ua_condition_ex"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/ua_discrete_alarm_ex"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/ua_exclusive_deviation_alarm_ex"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/ua_exclusive_limit_alarm_ex"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/ua_limit_alarm_ex"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/ua_non_exclusive_deviation_alarm_ex"), exports);
__exportStar(require("./interfaces/alarms_and_conditions/ua_non_exclusive_limit_alarm_ex"), exports);
var ua_multistate_discrete_impl_1 = require("../src/data_access/ua_multistate_discrete_impl");
Object.defineProperty(exports, "promoteToMultiStateDiscrete", { enumerable: true, get: function () { return ua_multistate_discrete_impl_1.promoteToMultiStateDiscrete; } });
var ua_multistate_value_discrete_impl_1 = require("../src/data_access/ua_multistate_value_discrete_impl");
Object.defineProperty(exports, "promoteToMultiStateValueDiscrete", { enumerable: true, get: function () { return ua_multistate_value_discrete_impl_1.promoteToMultiStateValueDiscrete; } });
var ua_two_state_discrete_impl_1 = require("../src/data_access/ua_two_state_discrete_impl");
Object.defineProperty(exports, "promoteToTwoStateDiscrete", { enumerable: true, get: function () { return ua_two_state_discrete_impl_1.promoteToTwoStateDiscrete; } });
__exportStar(require("./ua_root_folder"), exports);
__exportStar(require("./session_context"), exports);
__exportStar(require("./pseudo_session"), exports);
__exportStar(require("./helpers/dump_tools"), exports);
__exportStar(require("./helpers/make_optionals_map"), exports);
__exportStar(require("./helpers/argument_list"), exports);
__exportStar(require("./helpers/call_helpers"), exports);
__exportStar(require("./helpers/ensure_secure_access"), exports);
var adjust_browse_direction_1 = require("./helpers/adjust_browse_direction");
Object.defineProperty(exports, "adjustBrowseDirection", { enumerable: true, get: function () { return adjust_browse_direction_1.adjustBrowseDirection; } });
var resolve_opaque_on_address_space_1 = require("./helpers/resolve_opaque_on_address_space");
Object.defineProperty(exports, "resolveOpaqueOnAddressSpace", { enumerable: true, get: function () { return resolve_opaque_on_address_space_1.resolveOpaqueOnAddressSpace; } });
var continuation_point_manager_1 = require("./continuation_points/continuation_point_manager");
Object.defineProperty(exports, "ContinuationPointManager", { enumerable: true, get: function () { return continuation_point_manager_1.ContinuationPointManager; } });
var ensure_datatype_extracted_1 = require("./loader/ensure_datatype_extracted");
Object.defineProperty(exports, "ensureDatatypeExtracted", { enumerable: true, get: function () { return ensure_datatype_extracted_1.ensureDatatypeExtracted; } });
Object.defineProperty(exports, "ensureDatatypeExtractedWithCallback", { enumerable: true, get: function () { return ensure_datatype_extracted_1.ensureDatatypeExtractedWithCallback; } });
__exportStar(require("./loader/generateAddressSpaceRaw"), exports);
__exportStar(require("./namespace"), exports);
__exportStar(require("./namespace_data_access"), exports);
var ua_certificate_expiration_alarm_impl_1 = require("../src/alarms_and_conditions/ua_certificate_expiration_alarm_impl");
Object.defineProperty(exports, "instantiateCertificateExpirationAlarm", { enumerable: true, get: function () { return ua_certificate_expiration_alarm_impl_1.instantiateCertificateExpirationAlarm; } });
var nodeid_manager_1 = require("../src/nodeid_manager");
Object.defineProperty(exports, "NodeIdManager", { enumerable: true, get: function () { return nodeid_manager_1.NodeIdManager; } });
__exportStar(require("../src/event_data"), exports);
__exportStar(require("./set_namespace_meta_data"), exports);
__exportStar(require("node-opcua-address-space-base"), exports);
__exportStar(require("node-opcua-nodeset-ua"), exports);
