// node-opcua-service-filter/dist/make_content_filter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeContentFilter = exports.makeContentFilterElements = exports.inList = exports.greaterThan = exports.greaterThanOrEqual = exports.LessThanOrEqual = exports.lessThan = exports.and = exports.or = exports.s = exports.n = exports.l = exports.ofType = void 0;
/**
 * @module node-opcua-service-filter
 */
const node_opcua_basic_types_1 = require("node-opcua-basic-types");
const node_opcua_data_model_1 = require("node-opcua-data-model");
const node_opcua_nodeid_1 = require("node-opcua-nodeid");
const node_opcua_variant_1 = require("node-opcua-variant");
const imports_1 = require("./imports");
function ofType(nodeId) {
    const element = new imports_1.ContentFilterElement({
        filterOperator: imports_1.FilterOperator.OfType,
        filterOperands: [
            new imports_1.LiteralOperand({
                value: {
                    dataType: node_opcua_basic_types_1.DataType.NodeId,
                    value: (0, node_opcua_nodeid_1.resolveNodeId)(nodeId)
                }
            })
        ]
    });
    return element;
}
exports.ofType = ofType;
function l(dataType, value) {
    switch (dataType) {
        case node_opcua_basic_types_1.DataType.NodeId:
            value = (0, node_opcua_nodeid_1.resolveNodeId)(value);
    }
    return new imports_1.LiteralOperand({ value: new node_opcua_variant_1.Variant({ dataType, value }) });
}
exports.l = l;
function n(n) {
    return l(node_opcua_basic_types_1.DataType.NodeId, n);
}
exports.n = n;
function s(attributeId, path) {
    return new imports_1.SimpleAttributeOperand({
        attributeId: attributeId,
        browsePath: path.split(".").map(node_opcua_data_model_1.coerceQualifiedName)
    });
}
exports.s = s;
function or(a, b) {
    return new imports_1.ContentFilterElement({
        filterOperands: [a, b],
        filterOperator: imports_1.FilterOperator.Or
    });
}
exports.or = or;
function and(a, b) {
    return new imports_1.ContentFilterElement({
        filterOperands: [a, b],
        filterOperator: imports_1.FilterOperator.And
    });
}
exports.and = and;
function lessThan(a, b) {
    return new imports_1.ContentFilterElement({
        filterOperands: [a, b],
        filterOperator: imports_1.FilterOperator.LessThan
    });
}
exports.lessThan = lessThan;
function LessThanOrEqual(a, b) {
    return new imports_1.ContentFilterElement({
        filterOperands: [a, b],
        filterOperator: imports_1.FilterOperator.LessThanOrEqual
    });
}
exports.LessThanOrEqual = LessThanOrEqual;
function greaterThanOrEqual(a, b) {
    return new imports_1.ContentFilterElement({
        filterOperands: [a, b],
        filterOperator: imports_1.FilterOperator.GreaterThanOrEqual
    });
}
exports.greaterThanOrEqual = greaterThanOrEqual;
function greaterThan(a, b) {
    return new imports_1.ContentFilterElement({
        filterOperands: [a, b],
        filterOperator: imports_1.FilterOperator.GreaterThan
    });
}
exports.greaterThan = greaterThan;
function inList(a, l) {
    return new imports_1.ContentFilterElement({
        filterOperands: [a, ...l],
        filterOperator: imports_1.FilterOperator.InList
    });
}
exports.inList = inList;
function makeContentFilterElements(o) {
    const elements = [];
    function pushElement(element) {
        elements.push(element);
        const thisIndex = elements.length - 1;
        if (element.filterOperands) {
            for (let i = 0; i < element.filterOperands.length; i++) {
                const op = element.filterOperands[i];
                if (op instanceof imports_1.ContentFilterElement) {
                    const index = pushElement(op);
                    element.filterOperands[i] = new imports_1.ElementOperand({ index });
                }
            }
        }
        return thisIndex;
    }
    pushElement(o);
    return elements;
}
exports.makeContentFilterElements = makeContentFilterElements;
function makeContentFilter(o) {
    return new imports_1.ContentFilter({
        elements: makeContentFilterElements(o instanceof imports_1.ContentFilterElement ? o : new imports_1.ContentFilterElement(o))
    });
}
exports.makeContentFilter = makeContentFilter;
