// node-opcua-address-space/dist/src/state_machine/finite_state_machine.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.promoteToStateMachine = exports.UAStateMachineImpl = exports.UATransitionImpl = void 0;
/**
 * @module node-opcua-address-space
 */
const chalk = require("chalk");
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_constants_1 = require("node-opcua-constants");
const node_opcua_data_model_1 = require("node-opcua-data-model");
const node_opcua_data_model_2 = require("node-opcua-data-model");
const node_opcua_nodeid_1 = require("node-opcua-nodeid");
const node_opcua_status_code_1 = require("node-opcua-status-code");
const node_opcua_variant_1 = require("node-opcua-variant");
const register_node_promoter_1 = require("../../source/loader/register_node_promoter");
const ua_object_impl_1 = require("../ua_object_impl");
const base_node_impl_1 = require("../base_node_impl");
const doDebug = false;
class UATransitionImpl {
}
exports.UATransitionImpl = UATransitionImpl;
function getComponentFromTypeAndSubtype(typeDef) {
    const components_parts = [];
    components_parts.push(typeDef.getComponents());
    while (typeDef.subtypeOfObj) {
        typeDef = typeDef.subtypeOfObj;
        components_parts.push(typeDef.getComponents());
    }
    return Array.prototype.concat.apply([], components_parts).filter((x) => x.nodeClass === node_opcua_data_model_1.NodeClass.Object);
}
const defaultPredicate = (transitions, fromState, toState) => {
    if (transitions.length === 0) {
        return null;
    }
    if (transitions.length === 1) {
        return transitions[0];
    }
    // tslint:disable-next-line: no-console
    console.log(" FromState = ", fromState.browseName.toString());
    // tslint:disable-next-line: no-console
    console.log(" ToState   = ", toState.browseName.toString());
    for (const transition of transitions) {
        // tslint:disable-next-line: no-console
        console.log("  possible transition : ", transition.browseName.toString(), " ", transition.nodeId.toString());
    }
    // tslint:disable-next-line: no-console
    console.log("warning: a duplicated FromState Reference to the same target has been found.\nPlease check your model or provide a predicate method to select which one to use");
    // tslint:disable-next-line: no-console
    console.log("fromStateNode: ", fromState.toString());
    return transitions[0];
};
/*
 *
 * @class StateMachine
 * @constructor
 * @extends UAObject
 *
 *
 */
class UAStateMachineImpl extends ua_object_impl_1.UAObjectImpl {
    getStates() {
        const addressSpace = this.addressSpace;
        const initialStateType = addressSpace.findObjectType("InitialStateType");
        // istanbul ignore next
        if (!initialStateType) {
            throw new Error("cannot find InitialStateType");
        }
        const stateType = addressSpace.findObjectType("StateType");
        // istanbul ignore next
        if (!stateType) {
            throw new Error("cannot find StateType");
        }
        (0, node_opcua_assert_1.assert)(initialStateType.isSupertypeOf(stateType));
        const typeDef = this.typeDefinitionObj;
        let comp = getComponentFromTypeAndSubtype(typeDef);
        comp = comp.filter((c) => {
            if (!c.typeDefinitionObj || c.typeDefinitionObj.nodeClass !== node_opcua_data_model_1.NodeClass.ObjectType) {
                return false;
            }
            return c.typeDefinitionObj.isSupertypeOf(stateType);
        });
        return comp;
    }
    get states() {
        return this.getStates();
    }
    /**
     * @method getStateByName
     * @param name  the name of the state to get
     * @return the state with the given name
     */
    getStateByName(name) {
        let states = this.getStates();
        states = states.filter((s) => {
            return s.browseName.name === name;
        });
        (0, node_opcua_assert_1.assert)(states.length <= 1);
        return states.length === 1 ? states[0] : null;
    }
    getTransitions() {
        const addressSpace = this.addressSpace;
        const transitionType = addressSpace.findObjectType("TransitionType");
        // istanbul ignore next
        if (!transitionType) {
            throw new Error("cannot find TransitionType");
        }
        const typeDef = this.typeDefinitionObj;
        let comp = getComponentFromTypeAndSubtype(typeDef);
        comp = comp.filter((c) => {
            if (!c.typeDefinitionObj || c.typeDefinitionObj.nodeClass !== node_opcua_data_model_1.NodeClass.ObjectType) {
                return false;
            }
            return c.typeDefinitionObj.isSupertypeOf(transitionType);
        });
        return comp;
    }
    get transitions() {
        return this.getTransitions();
    }
    /**
     * return the node InitialStateType
     * @property initialState
     */
    get initialState() {
        const addressSpace = this.addressSpace;
        const initialStateType = addressSpace.findObjectType("InitialStateType");
        const typeDef = this.typeDefinitionObj;
        let comp = getComponentFromTypeAndSubtype(typeDef);
        comp = comp.filter((c) => c.typeDefinitionObj === initialStateType);
        // istanbul ignore next
        if (comp.length > 1) {
            throw new Error(" More than 1 initial state in stateMachine");
        }
        return comp.length === 0 ? null : comp[0];
    }
    /**
     *
     * @param node
     * @private
     */
    _coerceNode(node) {
        if (node === null) {
            return null;
        }
        const addressSpace = this.addressSpace;
        if (node instanceof base_node_impl_1.BaseNodeImpl) {
            return node;
        }
        else if (node instanceof node_opcua_nodeid_1.NodeId) {
            return addressSpace.findNode(node);
        }
        else if (typeof node === "string") {
            return this.getStateByName(node);
        }
        return null;
    }
    /**
     * @method isValidTransition
     * @param toStateNode
     * @return {boolean}
     */
    isValidTransition(toStateNode, predicate) {
        // is it legal to go from state currentState to toStateNode;
        if (!this.currentStateNode) {
            return true;
        }
        const n = this.currentState.readValue();
        // to be executed there must be a transition from currentState to toState
        const transition = this.findTransitionNode(this.currentStateNode, toStateNode, predicate);
        if (!transition) {
            // istanbul ignore next
            if (doDebug) {
                // tslint:disable-next-line: no-console
                console.log(" No transition from ", this.currentStateNode.browseName.toString(), " to ", toStateNode.toString());
            }
            return false;
        }
        return true;
    }
    /**
     */
    findTransitionNode(fromStateNode, toStateNode, predicate) {
        const addressSpace = this.addressSpace;
        const _fromStateNode = this._coerceNode(fromStateNode);
        if (!_fromStateNode) {
            return null;
        }
        const _toStateNode = this._coerceNode(toStateNode);
        if (!_toStateNode) {
            return null;
        }
        if (_fromStateNode.nodeClass !== node_opcua_data_model_1.NodeClass.Object) {
            throw new Error("Internal Error");
        }
        if (_toStateNode && _toStateNode.nodeClass !== node_opcua_data_model_1.NodeClass.Object) {
            throw new Error("Internal Error");
        }
        const stateType = addressSpace.findObjectType("StateType");
        if (!stateType) {
            throw new Error("Cannot find StateType");
        }
        (0, node_opcua_assert_1.assert)(_fromStateNode.typeDefinitionObj.isSupertypeOf(stateType));
        (0, node_opcua_assert_1.assert)(_toStateNode.typeDefinitionObj.isSupertypeOf(stateType));
        let transitions = _fromStateNode.findReferencesAsObject("FromState", false);
        transitions = transitions.filter((transition) => {
            (0, node_opcua_assert_1.assert)(transition.toStateNode.nodeClass === node_opcua_data_model_1.NodeClass.Object);
            return transition.toStateNode === _toStateNode;
        });
        if (transitions.length === 0) {
            // cannot find a transition from fromState to toState
            return null;
        }
        // istanbul ignore next
        if (transitions.length > 1) {
            const selectedTransition = (predicate || defaultPredicate)(transitions, _fromStateNode, _toStateNode);
            return selectedTransition;
        }
        return transitions[0];
    }
    get currentStateNode() {
        return this._currentStateNode;
    }
    /**
     * @property currentStateNode
     * @type BaseNode
     */
    set currentStateNode(value) {
        this._currentStateNode = value;
    }
    /**
     */
    getCurrentState() {
        // xx this.currentState.readValue().value.value.text
        // xx this.shelvingState.currentStateNode.browseName.toString()
        if (!this.currentStateNode) {
            return null;
        }
        return this.currentStateNode.browseName.toString();
    }
    /**
     * @method setState
     */
    setState(toStateNode, predicate) {
        if (!toStateNode) {
            this.currentStateNode = null;
            this.currentState.setValueFromSource({ dataType: node_opcua_variant_1.DataType.LocalizedText, value: null }, node_opcua_status_code_1.StatusCodes.BadStateNotActive);
            return;
        }
        if (typeof toStateNode === "string") {
            const state = this.getStateByName(toStateNode);
            // istanbul ignore next
            if (!state) {
                throw new Error("Cannot find state with name " + toStateNode);
            }
            (0, node_opcua_assert_1.assert)(state.browseName.name.toString() === toStateNode);
            toStateNode = state;
        }
        const fromStateNode = this.currentStateNode;
        toStateNode = this._coerceNode(toStateNode);
        (0, node_opcua_assert_1.assert)(toStateNode.nodeClass === node_opcua_data_model_1.NodeClass.Object);
        this.currentState.setValueFromSource({
            dataType: node_opcua_variant_1.DataType.LocalizedText,
            value: (0, node_opcua_data_model_1.coerceLocalizedText)(toStateNode.browseName.toString())
        }, node_opcua_status_code_1.StatusCodes.Good);
        this.currentStateNode = toStateNode;
        const transitionNode = this.findTransitionNode(fromStateNode, toStateNode, predicate);
        if (transitionNode) {
            // xx console.log("transitionNode ",transitionNode.toString());
            // The inherited Property SourceNode shall be filled with the NodeId of the StateMachine instance where the
            // Transition occurs. If the Transition occurs in a SubStateMachine, then the NodeId of the SubStateMachine
            // has to be used. If the Transition occurs between a StateMachine and a SubStateMachine, then the NodeId of
            // the StateMachine has to be used, independent of the direction of the Transition.
            // Transition identifies the Transition that triggered the Event.
            // FromState identifies the State before the Transition.
            // ToState identifies the State after the Transition.
            this.raiseEvent("TransitionEventType", {
                // Base EventType
                // xx nodeId:      this.nodeId,
                // TransitionEventType
                // TransitionVariableType
                transition: {
                    dataType: "LocalizedText",
                    value: transitionNode.displayName[0]
                },
                "transition.id": transitionNode.transitionNumber.readValue().value,
                fromState: {
                    dataType: "LocalizedText",
                    value: fromStateNode ? fromStateNode.displayName[0] : ""
                },
                "fromState.id": fromStateNode
                    ? fromStateNode.stateNumber.readValue().value
                    : {
                        dataType: "Null"
                    },
                toState: {
                    dataType: "LocalizedText",
                    value: toStateNode.displayName[0]
                },
                "toState.id": toStateNode.stateNumber.readValue().value
            });
        }
        else {
            if (fromStateNode && fromStateNode !== toStateNode) {
                // istanbul ignore next
                if (doDebug) {
                    const f = fromStateNode.browseName.toString();
                    const t = toStateNode.browseName.toString();
                    // tslint:disable-next-line:no-console
                    console.log(chalk.red("Warning"), " cannot raise event :  transition " + f + " to " + t + " is missing");
                }
            }
        }
        // also update executable flags on methods
        for (const method of this.getMethods()) {
            method._notifyAttributeChange(node_opcua_data_model_2.AttributeIds.Executable);
        }
    }
    /**
     * @internal
     * @private
     */
    _post_initialize() {
        const addressSpace = this.addressSpace;
        const finiteStateMachineType = addressSpace.findObjectType("FiniteStateMachineType");
        if (!finiteStateMachineType) {
            throw new Error("cannot find FiniteStateMachineType");
        }
        // xx assert(this.typeDefinitionObj && !this.subtypeOfObj);
        // xx assert(!this.typeDefinitionObj || this.typeDefinitionObj.isSupertypeOf(finiteStateMachineType));
        // get current Status
        const d = this.currentState.readValue();
        if (d.statusCode !== node_opcua_status_code_1.StatusCodes.Good) {
            this.setState(null);
        }
        else {
            const txt = d.value && d.value.value ? (d.value.value.text ? d.value.value.text.toString() : d.value.value.toString()) : "";
            this.currentStateNode = this.getStateByName(txt);
        }
    }
}
exports.UAStateMachineImpl = UAStateMachineImpl;
function promoteToStateMachine(node) {
    if (node instanceof UAStateMachineImpl) {
        return node; // already promoted
    }
    Object.setPrototypeOf(node, UAStateMachineImpl.prototype);
    (0, node_opcua_assert_1.assert)(node instanceof UAStateMachineImpl, "should now  be a State Machine");
    const _node = node;
    _node._post_initialize();
    return _node;
}
exports.promoteToStateMachine = promoteToStateMachine;
(0, register_node_promoter_1.registerNodePromoter)(node_opcua_constants_1.ObjectTypeIds.FiniteStateMachineType, promoteToStateMachine);
