// node-opcua-client/dist/alarms_and_conditions/client_tools.js
"use strict";
/**
 * @module node-opcua-client
 */
// tslint:disable:no-empty
// tslint:disable: only-arrow-functions
Object.defineProperty(exports, "__esModule", { value: true });
exports.callConditionRefresh = void 0;
const async = require("async");
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_data_model_1 = require("node-opcua-data-model");
const node_opcua_debug_1 = require("node-opcua-debug");
const node_opcua_nodeid_1 = require("node-opcua-nodeid");
const node_opcua_service_call_1 = require("node-opcua-service-call");
const node_opcua_service_translate_browse_path_1 = require("node-opcua-service-translate-browse-path");
const node_opcua_status_code_1 = require("node-opcua-status-code");
const node_opcua_variant_1 = require("node-opcua-variant");
const client_session_impl_1 = require("../private/client_session_impl");
const debugLog = (0, node_opcua_debug_1.make_debugLog)("A&E");
const doDebug = (0, node_opcua_debug_1.checkDebugFlag)("A&E");
const warningLog = (0, node_opcua_debug_1.make_warningLog)("A&E");
function callConditionRefresh(subscription, callback) {
    const subscriptionI = subscription;
    const theSession = subscriptionI.publishEngine.session;
    const subscriptionId = subscription.subscriptionId;
    (0, node_opcua_assert_1.assert)(isFinite(subscriptionId), "May be subscription is not yet initialized");
    (0, node_opcua_assert_1.assert)(typeof callback === "function");
    const conditionTypeNodeId = (0, node_opcua_nodeid_1.resolveNodeId)("ConditionType");
    let conditionRefreshId = (0, node_opcua_nodeid_1.resolveNodeId)("ConditionType_ConditionRefresh");
    async.series([
        // find conditionRefreshId
        (innerCallback) => {
            const browsePath = (0, node_opcua_service_translate_browse_path_1.makeBrowsePath)(conditionTypeNodeId, ".ConditionRefresh");
            theSession.translateBrowsePath(browsePath, (err, result) => {
                if (!err) {
                    result = result;
                    // istanbul ignore else
                    if (result.targets && result.targets.length > 0) {
                        conditionRefreshId = result.targets[0].targetId;
                    }
                    else {
                        // cannot find conditionRefreshId
                        debugLog("cannot find conditionRefreshId", result.toString());
                        err = new Error(" cannot find conditionRefreshId");
                    }
                }
                innerCallback(err ? err : undefined);
            });
        },
        (innerCallback) => {
            const methodToCall = {
                inputArguments: [new node_opcua_variant_1.Variant({ dataType: node_opcua_variant_1.DataType.UInt32, value: subscriptionId })],
                methodId: conditionRefreshId,
                objectId: conditionTypeNodeId
            };
            doDebug && debugLog("Calling method ", new node_opcua_service_call_1.CallMethodRequest(methodToCall).toString());
            theSession.call(methodToCall, (err, result) => {
                if (err) {
                    return innerCallback(err);
                }
                result = result;
                // istanbul ignore next
                if (result.statusCode !== node_opcua_status_code_1.StatusCodes.Good) {
                    warningLog(new node_opcua_service_call_1.CallMethodRequest(methodToCall).toString());
                    return innerCallback(new Error("Error " + result.statusCode.toString()));
                }
                innerCallback();
            });
        }
    ], (err) => {
        callback(err || undefined);
    });
}
exports.callConditionRefresh = callConditionRefresh;
client_session_impl_1.ClientSessionImpl.prototype.disableCondition = () => {
    /** */
};
client_session_impl_1.ClientSessionImpl.prototype.enableCondition = () => {
    /** */
};
client_session_impl_1.ClientSessionImpl.prototype.addCommentCondition = function (conditionId, eventId, comment, callback) {
    this._callMethodCondition("AddComment", conditionId, eventId, comment, callback);
};
client_session_impl_1.ClientSessionImpl.prototype.findMethodId = function (nodeId, methodName, callback) {
    const browsePath = (0, node_opcua_service_translate_browse_path_1.makeBrowsePath)(nodeId, "/" + methodName);
    let methodId;
    this.translateBrowsePath(browsePath, (err, result) => {
        if (err) {
            return callback(err);
        }
        /* istanbul ignore next */
        if (!result) {
            return callback(new Error("Internal Error"));
        }
        result.targets = result.targets || [];
        // istanbul ignore else
        if (result.targets.length > 0) {
            methodId = result.targets[0].targetId;
            return callback(null, methodId);
        }
        else {
            // cannot find objectWithMethodNodeId
            debugLog("cannot find " + methodName + " Method", result.toString());
            err = new Error(" cannot find " + methodName + " Method");
        }
        callback(err);
    });
};
/**
 *
 * @param methodName
 * @param conditionId
 * @param eventId
 * @param comment
 * @param callback
 * @private
 */
client_session_impl_1.ClientSessionImpl.prototype._callMethodCondition = function (methodName, conditionId, eventId, comment, callback) {
    conditionId = (0, node_opcua_nodeid_1.coerceNodeId)(conditionId);
    (0, node_opcua_assert_1.assert)(conditionId instanceof node_opcua_nodeid_1.NodeId);
    (0, node_opcua_assert_1.assert)(eventId instanceof Buffer);
    (0, node_opcua_assert_1.assert)(typeof comment === "string" || comment instanceof node_opcua_data_model_1.LocalizedText);
    comment = node_opcua_data_model_1.LocalizedText.coerce(comment) || new node_opcua_data_model_1.LocalizedText();
    let methodId;
    let statusCode;
    async.series([
        (innerCallback) => {
            this.findMethodId(conditionId, methodName, (err, _methodId) => {
                if (err) {
                    return innerCallback(err);
                }
                if (_methodId) {
                    methodId = _methodId;
                }
                innerCallback();
            });
        },
        (innerCallback) => {
            const methodToCalls = [];
            methodToCalls.push(new node_opcua_service_call_1.CallMethodRequest({
                inputArguments: [
                    /* eventId */ new node_opcua_variant_1.Variant({ dataType: "ByteString", value: eventId }),
                    /* comment */ new node_opcua_variant_1.Variant({ dataType: "LocalizedText", value: comment })
                ],
                methodId,
                objectId: conditionId
            }));
            this.call(methodToCalls, (err, results) => {
                if (err) {
                    return innerCallback(err);
                }
                statusCode = results[0].statusCode;
                innerCallback();
            });
        }
    ], (err) => {
        if (err) {
            return callback(err);
        }
        callback(null, statusCode);
    });
};
client_session_impl_1.ClientSessionImpl.prototype.confirmCondition = function (conditionId, eventId, comment, callback) {
    // ns=0;i=9113 AcknowledgeableConditionType#Confirm
    // note that confirm method is Optionals on condition
    this._callMethodCondition("Confirm", conditionId, eventId, comment, callback);
};
client_session_impl_1.ClientSessionImpl.prototype.acknowledgeCondition = function (conditionId, eventId, comment, callback) {
    // ns=0;i=9111 AcknowledgeableConditionType#Acknowledge
    this._callMethodCondition("Acknowledge", conditionId, eventId, comment, callback);
};
// tslint:disable:no-var-requires
// tslint:disable:max-line-length
const thenify = require("thenify");
const opts = { multiArgs: false };
client_session_impl_1.ClientSessionImpl.prototype.addCommentCondition = thenify.withCallback(client_session_impl_1.ClientSessionImpl.prototype.addCommentCondition, opts);
client_session_impl_1.ClientSessionImpl.prototype.findMethodId = thenify.withCallback(client_session_impl_1.ClientSessionImpl.prototype.findMethodId, opts);
client_session_impl_1.ClientSessionImpl.prototype.confirmCondition = thenify.withCallback(client_session_impl_1.ClientSessionImpl.prototype.confirmCondition, opts);
client_session_impl_1.ClientSessionImpl.prototype.acknowledgeCondition = thenify.withCallback(client_session_impl_1.ClientSessionImpl.prototype.acknowledgeCondition, opts);
module.exports.callConditionRefresh = thenify.withCallback(module.exports.callConditionRefresh, opts);
