// node-opcua-client-proxy/dist/proxy_base_node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProxyBaseNode = void 0;
/**
 * @module node-opcua-client-proxy
 */
const events_1 = require("events");
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_data_model_1 = require("node-opcua-data-model");
const node_opcua_status_code_1 = require("node-opcua-status-code");
class ProxyBaseNode extends events_1.EventEmitter {
    constructor(proxyManager, nodeId, nodeClass) {
        super();
        this.nodeId = nodeId;
        this.proxyManager = proxyManager;
        (0, node_opcua_assert_1.assert)(this.proxyManager.session, "expecting valid session");
        Object.defineProperty(this, "proxyManager", {
            enumerable: false,
            writable: true
        });
        this.$components = [];
        this.$properties = [];
        this.$methods = [];
        this.$organizes = [];
        this.description = "";
        this.browseName = "";
        this.nodeClass = nodeClass;
    }
    /**
     * get a updated Value of the Variable , by using a ReadRequest
     */
    readValue(callback) {
        (0, node_opcua_assert_1.assert)(this.proxyManager);
        const session = this.proxyManager.session;
        (0, node_opcua_assert_1.assert)(session);
        const nodeToRead = {
            attributeId: node_opcua_data_model_1.AttributeIds.Value,
            nodeId: this.nodeId
        };
        this.proxyManager.session.read(nodeToRead, (err, dataValue) => {
            // istanbul ignore next
            if (err) {
                return callback(err);
            }
            const data = dataValue.value;
            callback(null, data);
        });
    }
    /**
     * set the Value of the Variable, by using a WriteRequest
     */
    writeValue(dataValue, callback) {
        (0, node_opcua_assert_1.assert)(this.proxyManager);
        const session = this.proxyManager.session;
        (0, node_opcua_assert_1.assert)(session);
        const nodeToWrite = {
            attributeId: node_opcua_data_model_1.AttributeIds.Value,
            nodeId: this.nodeId,
            value: dataValue
        };
        this.proxyManager.session.write(nodeToWrite, (err, statusCode) => {
            // istanbul ignore next
            if (err) {
                return callback(err);
            }
            if (statusCode !== node_opcua_status_code_1.StatusCodes.Good) {
                callback(new Error(statusCode.toString()));
            }
            else {
                callback();
            }
        });
    }
    toString() {
        const str = [];
        str.push(" ProxyObject ");
        str.push("   browseName     : " + this.browseName.toString());
        // str.push("   typeDefinition : " + this.typeDefinition.toString());
        str.push("   $components#   : " + this.$components.length.toString());
        str.push("   $properties#   : " + this.$properties.length.toString());
        return str.join("\n");
    }
}
exports.ProxyBaseNode = ProxyBaseNode;
// tslint:disable:no-var-requires
const thenify = require("thenify");
ProxyBaseNode.prototype.readValue = thenify.withCallback(ProxyBaseNode.prototype.readValue);
ProxyBaseNode.prototype.writeValue = thenify.withCallback(ProxyBaseNode.prototype.writeValue);
