// node-opcua-service-translate-browse-path/dist/tools_browse_path.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.constructBrowsePathFromQualifiedName = exports.stringToQualifiedName = void 0;
/**
 * @module node-opcua-service-translate-browse-path
 * @class ToolBrowsePath
 * @static
 */
const node_opcua_constants_1 = require("node-opcua-constants");
const node_opcua_nodeid_1 = require("node-opcua-nodeid");
const imports_1 = require("./imports");
const hierarchicalReferencesId = (0, node_opcua_nodeid_1.makeNodeId)(node_opcua_constants_1.ReferenceTypeIds.HierarchicalReferences);
var node_opcua_data_model_1 = require("node-opcua-data-model");
Object.defineProperty(exports, "stringToQualifiedName", { enumerable: true, get: function () { return node_opcua_data_model_1.stringToQualifiedName; } });
/**
 * @method constructBrowsePathFromQualifiedName
 * @param startingNode
 * @param targetNames
 * @return {BrowsePath}
 */
function constructBrowsePathFromQualifiedName(startingNode, targetNames) {
    targetNames = targetNames || [];
    const elements = targetNames.map((targetName) => {
        return {
            isInverse: false,
            includeSubtypes: true,
            referenceTypeId: hierarchicalReferencesId,
            targetName
        };
    });
    const browsePath = new imports_1.BrowsePath({
        relativePath: { elements },
        startingNode: startingNode.nodeId, // ROOT
    });
    return browsePath;
}
exports.constructBrowsePathFromQualifiedName = constructBrowsePathFromQualifiedName;
