// node-opcua-nodeset-ua/dist/enum_exception_deviation_format.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumExceptionDeviationFormat = void 0;
/**
 * |           |                                                  |
 * |-----------|--------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                      |
 * | nodeClass |DataType                                          |
 * | name      |ExceptionDeviationFormat                          |
 * | isAbstract|false                                             |
 */
var EnumExceptionDeviationFormat;
(function (EnumExceptionDeviationFormat) {
    EnumExceptionDeviationFormat[EnumExceptionDeviationFormat["AbsoluteValue"] = 0] = "AbsoluteValue";
    EnumExceptionDeviationFormat[EnumExceptionDeviationFormat["PercentOfValue"] = 1] = "PercentOfValue";
    EnumExceptionDeviationFormat[EnumExceptionDeviationFormat["PercentOfRange"] = 2] = "PercentOfRange";
    EnumExceptionDeviationFormat[EnumExceptionDeviationFormat["PercentOfEURange"] = 3] = "PercentOfEURange";
    EnumExceptionDeviationFormat[EnumExceptionDeviationFormat["Unknown"] = 4] = "Unknown";
})(EnumExceptionDeviationFormat = exports.EnumExceptionDeviationFormat || (exports.EnumExceptionDeviationFormat = {}));
