// node-opcua-nodeset-ua/dist/enum_structure.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumStructure = void 0;
/**
 * |           |                                                  |
 * |-----------|--------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                      |
 * | nodeClass |DataType                                          |
 * | name      |StructureType                                     |
 * | isAbstract|false                                             |
 */
var EnumStructure;
(function (EnumStructure) {
    EnumStructure[EnumStructure["Structure"] = 0] = "Structure";
    EnumStructure[EnumStructure["StructureWithOptionalFields"] = 1] = "StructureWithOptionalFields";
    EnumStructure[EnumStructure["Union"] = 2] = "Union";
    EnumStructure[EnumStructure["StructureWithSubtypedValues"] = 3] = "StructureWithSubtypedValues";
    EnumStructure[EnumStructure["UnionWithSubtypedValues"] = 4] = "UnionWithSubtypedValues";
})(EnumStructure = exports.EnumStructure || (exports.EnumStructure = {}));
