// node-opcua-nodeset-ua/dist/enum_node_attributes_mask.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumNodeAttributesMask = void 0;
/**
 * |           |                                                  |
 * |-----------|--------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                      |
 * | nodeClass |DataType                                          |
 * | name      |NodeAttributesMask                                |
 * | isAbstract|false                                             |
 */
var EnumNodeAttributesMask;
(function (EnumNodeAttributesMask) {
    EnumNodeAttributesMask[EnumNodeAttributesMask["None"] = 0] = "None";
    EnumNodeAttributesMask[EnumNodeAttributesMask["AccessLevel"] = 1] = "AccessLevel";
    EnumNodeAttributesMask[EnumNodeAttributesMask["ArrayDimensions"] = 2] = "ArrayDimensions";
    EnumNodeAttributesMask[EnumNodeAttributesMask["BrowseName"] = 4] = "BrowseName";
    EnumNodeAttributesMask[EnumNodeAttributesMask["ContainsNoLoops"] = 8] = "ContainsNoLoops";
    EnumNodeAttributesMask[EnumNodeAttributesMask["DataType"] = 16] = "DataType";
    EnumNodeAttributesMask[EnumNodeAttributesMask["Description"] = 32] = "Description";
    EnumNodeAttributesMask[EnumNodeAttributesMask["DisplayName"] = 64] = "DisplayName";
    EnumNodeAttributesMask[EnumNodeAttributesMask["EventNotifier"] = 128] = "EventNotifier";
    EnumNodeAttributesMask[EnumNodeAttributesMask["Executable"] = 256] = "Executable";
    EnumNodeAttributesMask[EnumNodeAttributesMask["Historizing"] = 512] = "Historizing";
    EnumNodeAttributesMask[EnumNodeAttributesMask["InverseName"] = 1024] = "InverseName";
    EnumNodeAttributesMask[EnumNodeAttributesMask["IsAbstract"] = 2048] = "IsAbstract";
    EnumNodeAttributesMask[EnumNodeAttributesMask["MinimumSamplingInterval"] = 4096] = "MinimumSamplingInterval";
    EnumNodeAttributesMask[EnumNodeAttributesMask["NodeClass"] = 8192] = "NodeClass";
    EnumNodeAttributesMask[EnumNodeAttributesMask["NodeId"] = 16384] = "NodeId";
    EnumNodeAttributesMask[EnumNodeAttributesMask["Symmetric"] = 32768] = "Symmetric";
    EnumNodeAttributesMask[EnumNodeAttributesMask["UserAccessLevel"] = 65536] = "UserAccessLevel";
    EnumNodeAttributesMask[EnumNodeAttributesMask["UserExecutable"] = 131072] = "UserExecutable";
    EnumNodeAttributesMask[EnumNodeAttributesMask["UserWriteMask"] = 262144] = "UserWriteMask";
    EnumNodeAttributesMask[EnumNodeAttributesMask["ValueRank"] = 524288] = "ValueRank";
    EnumNodeAttributesMask[EnumNodeAttributesMask["WriteMask"] = 1048576] = "WriteMask";
    EnumNodeAttributesMask[EnumNodeAttributesMask["Value"] = 2097152] = "Value";
    EnumNodeAttributesMask[EnumNodeAttributesMask["DataTypeDefinition"] = 4194304] = "DataTypeDefinition";
    EnumNodeAttributesMask[EnumNodeAttributesMask["RolePermissions"] = 8388608] = "RolePermissions";
    EnumNodeAttributesMask[EnumNodeAttributesMask["AccessRestrictions"] = 16777216] = "AccessRestrictions";
    EnumNodeAttributesMask[EnumNodeAttributesMask["All"] = 33554431] = "All";
    EnumNodeAttributesMask[EnumNodeAttributesMask["BaseNode"] = 26501220] = "BaseNode";
    EnumNodeAttributesMask[EnumNodeAttributesMask["Object"] = 26501348] = "Object";
    EnumNodeAttributesMask[EnumNodeAttributesMask["ObjectType"] = 26503268] = "ObjectType";
    EnumNodeAttributesMask[EnumNodeAttributesMask["Variable"] = 26571383] = "Variable";
    EnumNodeAttributesMask[EnumNodeAttributesMask["VariableType"] = 28600438] = "VariableType";
    EnumNodeAttributesMask[EnumNodeAttributesMask["Method"] = 26632548] = "Method";
    EnumNodeAttributesMask[EnumNodeAttributesMask["ReferenceType"] = 26537060] = "ReferenceType";
    EnumNodeAttributesMask[EnumNodeAttributesMask["View"] = 26501356] = "View";
})(EnumNodeAttributesMask = exports.EnumNodeAttributesMask || (exports.EnumNodeAttributesMask = {}));
