// node-opcua-nodeset-ua/dist/enum_pub_sub_state.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumPubSubState = void 0;
/**
 * |           |                                                  |
 * |-----------|--------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                      |
 * | nodeClass |DataType                                          |
 * | name      |PubSubState                                       |
 * | isAbstract|false                                             |
 */
var EnumPubSubState;
(function (EnumPubSubState) {
    EnumPubSubState[EnumPubSubState["Disabled"] = 0] = "Disabled";
    EnumPubSubState[EnumPubSubState["Paused"] = 1] = "Paused";
    EnumPubSubState[EnumPubSubState["Operational"] = 2] = "Operational";
    EnumPubSubState[EnumPubSubState["Error"] = 3] = "Error";
    EnumPubSubState[EnumPubSubState["PreOperational"] = 4] = "PreOperational";
})(EnumPubSubState = exports.EnumPubSubState || (exports.EnumPubSubState = {}));
