// node-opcua-nodeset-ua/dist/enum_identity_criteria.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumIdentityCriteria = void 0;
/**
 * |           |                                                  |
 * |-----------|--------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                      |
 * | nodeClass |DataType                                          |
 * | name      |IdentityCriteriaType                              |
 * | isAbstract|false                                             |
 */
var EnumIdentityCriteria;
(function (EnumIdentityCriteria) {
    /**
     * The rule specifies a UserName from a
     * UserNameIdentityToken.
     */
    EnumIdentityCriteria[EnumIdentityCriteria["UserName"] = 1] = "UserName";
    /**
     * The rule specifies the Thumbprint of a user or CA
     * Certificate.
     */
    EnumIdentityCriteria[EnumIdentityCriteria["Thumbprint"] = 2] = "Thumbprint";
    /**
     * The rule is a Role specified in an Access Token.
     */
    EnumIdentityCriteria[EnumIdentityCriteria["Role"] = 3] = "Role";
    /**
     * The rule is a user group specified in the Access
     * Token.
     */
    EnumIdentityCriteria[EnumIdentityCriteria["GroupId"] = 4] = "GroupId";
    /**
     * The rule specifies Anonymous UserIdentityToken.
     */
    EnumIdentityCriteria[EnumIdentityCriteria["Anonymous"] = 5] = "Anonymous";
    /**
     * The rule specifies any non Anonymous
     * UserIdentityToken.
     */
    EnumIdentityCriteria[EnumIdentityCriteria["AuthenticatedUser"] = 6] = "AuthenticatedUser";
    /**
     * The rule specifies the combination of an
     * application identity and an Anonymous
     * UserIdentityToken.
     */
    EnumIdentityCriteria[EnumIdentityCriteria["Application"] = 7] = "Application";
    /**
     * The rule specifies the X509 subject name of a
     * user or CA Certificate.
     */
    EnumIdentityCriteria[EnumIdentityCriteria["X509Subject"] = 8] = "X509Subject";
})(EnumIdentityCriteria = exports.EnumIdentityCriteria || (exports.EnumIdentityCriteria = {}));
