// node-opcua-nodeset-ua/dist/enum_diagnostics_level.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumDiagnosticsLevel = void 0;
/**
 * |           |                                                  |
 * |-----------|--------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                      |
 * | nodeClass |DataType                                          |
 * | name      |DiagnosticsLevel                                  |
 * | isAbstract|false                                             |
 */
var EnumDiagnosticsLevel;
(function (EnumDiagnosticsLevel) {
    EnumDiagnosticsLevel[EnumDiagnosticsLevel["Basic"] = 0] = "Basic";
    EnumDiagnosticsLevel[EnumDiagnosticsLevel["Advanced"] = 1] = "Advanced";
    EnumDiagnosticsLevel[EnumDiagnosticsLevel["Info"] = 2] = "Info";
    EnumDiagnosticsLevel[EnumDiagnosticsLevel["Log"] = 3] = "Log";
    EnumDiagnosticsLevel[EnumDiagnosticsLevel["Debug"] = 4] = "Debug";
})(EnumDiagnosticsLevel = exports.EnumDiagnosticsLevel || (exports.EnumDiagnosticsLevel = {}));
