// node-opcua-nodeset-ua/dist/enum_broker_transport_quality_of_service.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumBrokerTransportQualityOfService = void 0;
/**
 * |           |                                                  |
 * |-----------|--------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                      |
 * | nodeClass |DataType                                          |
 * | name      |BrokerTransportQualityOfService                   |
 * | isAbstract|false                                             |
 */
var EnumBrokerTransportQualityOfService;
(function (EnumBrokerTransportQualityOfService) {
    EnumBrokerTransportQualityOfService[EnumBrokerTransportQualityOfService["NotSpecified"] = 0] = "NotSpecified";
    EnumBrokerTransportQualityOfService[EnumBrokerTransportQualityOfService["BestEffort"] = 1] = "BestEffort";
    EnumBrokerTransportQualityOfService[EnumBrokerTransportQualityOfService["AtLeastOnce"] = 2] = "AtLeastOnce";
    EnumBrokerTransportQualityOfService[EnumBrokerTransportQualityOfService["AtMostOnce"] = 3] = "AtMostOnce";
    EnumBrokerTransportQualityOfService[EnumBrokerTransportQualityOfService["ExactlyOnce"] = 4] = "ExactlyOnce";
})(EnumBrokerTransportQualityOfService = exports.EnumBrokerTransportQualityOfService || (exports.EnumBrokerTransportQualityOfService = {}));
