// node-opcua-nodeset-ua/dist/enum_server_state.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumServerState = void 0;
/**
 * |           |                                                  |
 * |-----------|--------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                      |
 * | nodeClass |DataType                                          |
 * | name      |ServerState                                       |
 * | isAbstract|false                                             |
 */
var EnumServerState;
(function (EnumServerState) {
    EnumServerState[EnumServerState["Running"] = 0] = "Running";
    EnumServerState[EnumServerState["Failed"] = 1] = "Failed";
    EnumServerState[EnumServerState["NoConfiguration"] = 2] = "NoConfiguration";
    EnumServerState[EnumServerState["Suspended"] = 3] = "Suspended";
    EnumServerState[EnumServerState["Shutdown"] = 4] = "Shutdown";
    EnumServerState[EnumServerState["Test"] = 5] = "Test";
    EnumServerState[EnumServerState["CommunicationFault"] = 6] = "CommunicationFault";
    EnumServerState[EnumServerState["Unknown"] = 7] = "Unknown";
})(EnumServerState = exports.EnumServerState || (exports.EnumServerState = {}));
