// node-opcua-client/dist/alarms_and_conditions/client_alarm_tools_extractConditionFields.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractConditionFields = void 0;
const node_opcua_data_model_1 = require("node-opcua-data-model");
const node_opcua_nodeid_1 = require("node-opcua-nodeid");
const node_opcua_types_1 = require("node-opcua-types");
function extractConditionFields(session, conditionNodeId) {
    return __awaiter(this, void 0, void 0, function* () {
        // conditionNodeId could be a Object of type ConditionType
        // or it could be directly a ObjectType which is a subType of ConditionType
        const _duplicateMap = {};
        const fields1 = [];
        function addField(name) {
            if (!_duplicateMap[name]) {
                fields1.push(name);
                _duplicateMap[name] = name;
            }
        }
        const stack = [];
        function deferObjectOrVariableInvestigation(objectId, browseName) {
            stack.push({
                browseName,
                nodeId: objectId
            });
        }
        function _investigateObjectOrVariable() {
            return __awaiter(this, void 0, void 0, function* () {
                if (stack.length === 0) {
                    return;
                }
                const extracted = stack.splice(0);
                const nodesToBrowse = extracted.map((e) => {
                    const b = {
                        browseDirection: node_opcua_data_model_1.BrowseDirection.Forward,
                        includeSubtypes: true,
                        // tslint:disable-next-line: no-bitwise
                        nodeClassMask: node_opcua_data_model_1.NodeClassMask.Object | node_opcua_data_model_1.NodeClassMask.Variable,
                        nodeId: e.nodeId,
                        referenceTypeId: "HasChild",
                        resultMask: 63
                    };
                    return b;
                });
                const results = yield session.browse(nodesToBrowse);
                let i = 0;
                for (const result of results) {
                    const name = extracted[i].browseName.toString();
                    i++;
                    if (!result.references) {
                        continue;
                    }
                    for (const ref of result.references) {
                        const n = name + "." + ref.browseName.toString();
                        if (ref.nodeClass === node_opcua_types_1.NodeClass.Variable) {
                            addField(n);
                        }
                        deferObjectOrVariableInvestigation(ref.nodeId, n);
                    }
                }
                yield _investigateObjectOrVariable();
            });
        }
        // tslint:disable-next-line: no-shadowed-variable
        function _investigateLevel(conditionNodeId) {
            return __awaiter(this, void 0, void 0, function* () {
                const nodeToBrowse1 = {
                    browseDirection: node_opcua_data_model_1.BrowseDirection.Inverse,
                    includeSubtypes: true,
                    nodeClassMask: node_opcua_data_model_1.NodeClassMask.ObjectType,
                    nodeId: conditionNodeId,
                    referenceTypeId: (0, node_opcua_nodeid_1.resolveNodeId)("HasSubtype"),
                    resultMask: 63
                };
                const nodeToBrowse2 = {
                    browseDirection: node_opcua_data_model_1.BrowseDirection.Forward,
                    includeSubtypes: true,
                    // tslint:disable-next-line: no-bitwise
                    nodeClassMask: node_opcua_data_model_1.NodeClassMask.Object | node_opcua_data_model_1.NodeClassMask.Variable,
                    nodeId: conditionNodeId,
                    referenceTypeId: (0, node_opcua_nodeid_1.resolveNodeId)("HasChild"),
                    resultMask: 63
                };
                const nodesToBrowse = [nodeToBrowse1, nodeToBrowse2];
                const browseResults = yield session.browse(nodesToBrowse);
                if (browseResults[1] && browseResults[1].references) {
                    for (const ref of browseResults[1].references) {
                        if (ref.nodeClass === node_opcua_types_1.NodeClass.Variable) {
                            addField(ref.browseName.toString());
                        }
                        deferObjectOrVariableInvestigation(ref.nodeId, ref.browseName.toString());
                    }
                }
                if (browseResults[0] && browseResults[0].references) {
                    const promises = [];
                    for (const reference of browseResults[0].references) {
                        promises.push(_investigateLevel(reference.nodeId));
                    }
                    yield Promise.all(promises);
                }
                yield _investigateObjectOrVariable();
            });
        }
        yield _investigateLevel(conditionNodeId);
        // add this field which will always be added
        addField("ConditionId");
        return fields1;
    });
}
exports.extractConditionFields = extractConditionFields;
