// node-opcua-basic-types/dist/string.js
"use strict";
/***
 * @module node-opcua-basic-types
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.encodeUAString = exports.decodeUAString = exports.encodeString = exports.decodeString = exports.randomString = exports.isValidString = void 0;
const utils_1 = require("./utils");
function isValidString(value) {
    return typeof value === "string";
}
exports.isValidString = isValidString;
function randomString() {
    const nbCar = (0, utils_1.getRandomInt)(1, 20);
    const cars = [];
    for (let i = 0; i < nbCar; i++) {
        cars.push(String.fromCharCode(65 + (0, utils_1.getRandomInt)(0, 26)));
    }
    return cars.join("");
}
exports.randomString = randomString;
function decodeString(stream, value) {
    return stream.readString();
}
exports.decodeString = decodeString;
function encodeString(value, stream) {
    stream.writeString(value);
}
exports.encodeString = encodeString;
exports.decodeUAString = decodeString;
exports.encodeUAString = encodeString;
