// node-opcua-constants/dist/aggregate_function_enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregateFunction = void 0;
var AggregateFunction;
(function (AggregateFunction) {
    AggregateFunction[AggregateFunction["AnnotationCount"] = 2351] = "AnnotationCount";
    AggregateFunction[AggregateFunction["Average"] = 2342] = "Average";
    AggregateFunction[AggregateFunction["Count"] = 2352] = "Count";
    AggregateFunction[AggregateFunction["Delta"] = 2359] = "Delta";
    AggregateFunction[AggregateFunction["DeltaBounds"] = 11507] = "DeltaBounds";
    AggregateFunction[AggregateFunction["DurationBad"] = 2361] = "DurationBad";
    AggregateFunction[AggregateFunction["DurationGood"] = 2360] = "DurationGood";
    AggregateFunction[AggregateFunction["DurationInStateNonZero"] = 11308] = "DurationInStateNonZero";
    AggregateFunction[AggregateFunction["DurationInStateZero"] = 2358] = "DurationInStateZero";
    AggregateFunction[AggregateFunction["EndBound"] = 11506] = "EndBound";
    AggregateFunction[AggregateFunction["Interpolative"] = 2341] = "Interpolative";
    AggregateFunction[AggregateFunction["Maximum"] = 2347] = "Maximum";
    AggregateFunction[AggregateFunction["Maximum2"] = 11287] = "Maximum2";
    AggregateFunction[AggregateFunction["MaximumActualTime"] = 2349] = "MaximumActualTime";
    AggregateFunction[AggregateFunction["MaximumActualTime2"] = 11306] = "MaximumActualTime2";
    AggregateFunction[AggregateFunction["Minimum"] = 2346] = "Minimum";
    AggregateFunction[AggregateFunction["Minimum2"] = 11286] = "Minimum2";
    AggregateFunction[AggregateFunction["MinimumActualTime"] = 2348] = "MinimumActualTime";
    AggregateFunction[AggregateFunction["MinimumActualTime2"] = 11305] = "MinimumActualTime2";
    AggregateFunction[AggregateFunction["NumberOfTransitions"] = 2355] = "NumberOfTransitions";
    AggregateFunction[AggregateFunction["PercentBad"] = 2363] = "PercentBad";
    AggregateFunction[AggregateFunction["PercentGood"] = 2362] = "PercentGood";
    AggregateFunction[AggregateFunction["Range"] = 2350] = "Range";
    AggregateFunction[AggregateFunction["Range2"] = 11288] = "Range2";
    AggregateFunction[AggregateFunction["StandardDeviationPopulation"] = 11427] = "StandardDeviationPopulation";
    AggregateFunction[AggregateFunction["StandardDeviationSample"] = 11426] = "StandardDeviationSample";
    AggregateFunction[AggregateFunction["Start"] = 2357] = "Start";
    AggregateFunction[AggregateFunction["StartBound"] = 11505] = "StartBound";
    AggregateFunction[AggregateFunction["TimeAverage"] = 2343] = "TimeAverage";
    AggregateFunction[AggregateFunction["TimeAverage2"] = 11285] = "TimeAverage2";
    AggregateFunction[AggregateFunction["Total"] = 2344] = "Total";
    AggregateFunction[AggregateFunction["Total2"] = 11304] = "Total2";
    AggregateFunction[AggregateFunction["VariancePopulation"] = 11429] = "VariancePopulation";
    AggregateFunction[AggregateFunction["VarianceSample"] = 11428] = "VarianceSample";
    AggregateFunction[AggregateFunction["WorstQuality"] = 2364] = "WorstQuality";
    AggregateFunction[AggregateFunction["WorstQuality2"] = 11292] = "WorstQuality2";
})(AggregateFunction = exports.AggregateFunction || (exports.AggregateFunction = {}));
