// node-opcua-factory/dist/encode_decode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decodeNull = exports.encodeNull = exports.decodeAny = exports.encodeAny = exports.toJSONGuid = exports.defaultGuidValue = exports.defaultDecode = exports.defaultEncode = void 0;
const node_opcua_assert_1 = require("node-opcua-assert");
function defaultEncode(value, stream) {
    /** */
}
exports.defaultEncode = defaultEncode;
function defaultDecode(stream) {
    return null;
}
exports.defaultDecode = defaultDecode;
function defaultGuidValue() {
    return Buffer.alloc(0);
}
exports.defaultGuidValue = defaultGuidValue;
function toJSONGuid(value) {
    if (typeof value === "string") {
        return value;
    }
    (0, node_opcua_assert_1.default)(value instanceof Buffer);
    return value.toString("base64");
}
exports.toJSONGuid = toJSONGuid;
function encodeAny(value, stream) {
    (0, node_opcua_assert_1.default)(false, "type 'Any' cannot be encoded");
}
exports.encodeAny = encodeAny;
function decodeAny(stream) {
    (0, node_opcua_assert_1.default)(false, "type 'Any' cannot be decoded");
}
exports.decodeAny = decodeAny;
// eslint-disable-next-line @typescript-eslint/no-empty-function
function encodeNull(value, stream) { }
exports.encodeNull = encodeNull;
function decodeNull(stream) {
    return null;
}
exports.decodeNull = decodeNull;
