// node-opcua-server/dist/register_server_manager_mdns_only.js
"use strict";
/**
 * @module node-opcua-server
 */
// RegisterServerManagerMDNSONLY
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegisterServerManagerMDNSONLY = void 0;
const events_1 = require("events");
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_service_discovery_1 = require("node-opcua-service-discovery");
const base_server_1 = require("./base_server");
/**
 * a RegisterServerManager that declare the server the OPCUA Bonjour service
 * available on the current computer
 */
class RegisterServerManagerMDNSONLY extends events_1.EventEmitter {
    constructor(options) {
        super();
        this.discoveryServerEndpointUrl = "";
        this.server = options.server;
        (0, node_opcua_assert_1.assert)(this.server);
        (0, node_opcua_assert_1.assert)(this.server instanceof base_server_1.OPCUABaseServer);
        this.bonjour = new node_opcua_service_discovery_1.BonjourHolder();
    }
    stop(callback) {
        if (this.bonjour) {
            this.bonjour.stopAnnouncedOnMulticastSubnetWithCallback(() => {
                this.emit("serverUnregistered");
                setImmediate(callback);
            });
        }
    }
    start(callback) {
        // istanbul ignore next
        if (!this.server) {
            throw new Error("internal error");
        }
        (0, node_opcua_assert_1.assert)(this.server instanceof base_server_1.OPCUABaseServer);
        this.bonjour.announcedOnMulticastSubnetWithCallback({
            capabilities: this.server.capabilitiesForMDNS,
            name: this.server.serverInfo.applicationUri,
            path: "/",
            port: this.server.endpoints[0].port
        }, () => {
            this.emit("serverRegistered");
            setImmediate(callback);
        });
    }
    dispose() {
        (0, node_opcua_assert_1.assert)(!this.bonjour.isStarted());
        (0, node_opcua_assert_1.assert)(this.server);
        this.server = undefined;
    }
}
exports.RegisterServerManagerMDNSONLY = RegisterServerManagerMDNSONLY;
