// node-opcua-address-space/dist/src/index_current.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeIdManager = exports.VariableHistorian = exports.resolveReferenceType = exports.resolveReferenceNode = exports.makeAttributeEventName = exports.adjustNamespaceArray = exports.dumpToBSD = exports.dumpXml = exports.NamespaceOptions = exports.promoteToStateMachine = exports.ContinuationPointManager = exports.ensureDatatypeExtractedWithCallback = exports.ensureDatatypeExtracted = exports.AddressSpace = void 0;
/**
 * @module node-opcua-address-space
 */
var address_space_1 = require("./address_space");
Object.defineProperty(exports, "AddressSpace", { enumerable: true, get: function () { return address_space_1.AddressSpace; } });
__exportStar(require("../source/session_context"), exports);
__exportStar(require("../source/helpers/dump_tools"), exports);
__exportStar(require("../source/helpers/adjust_browse_direction"), exports);
__exportStar(require("../source/pseudo_session"), exports);
__exportStar(require("../source/helpers/make_optionals_map"), exports);
__exportStar(require("../source/helpers/argument_list"), exports);
__exportStar(require("../source/helpers/call_helpers"), exports);
__exportStar(require("../source/helpers/ensure_secure_access"), exports);
__exportStar(require("../source/helpers/resolve_opaque_on_address_space"), exports);
__exportStar(require("../source/interfaces/alarms_and_conditions/condition_info_i"), exports);
__exportStar(require("../source/set_namespace_meta_data"), exports);
__exportStar(require("../source/namespace"), exports);
__exportStar(require("../source/namespace_data_access"), exports);
__exportStar(require("../source/namespace_machine_state"), exports);
__exportStar(require("../source/namespace_alarm_and_condition"), exports);
var ensure_datatype_extracted_1 = require("../source/loader/ensure_datatype_extracted");
Object.defineProperty(exports, "ensureDatatypeExtracted", { enumerable: true, get: function () { return ensure_datatype_extracted_1.ensureDatatypeExtracted; } });
Object.defineProperty(exports, "ensureDatatypeExtractedWithCallback", { enumerable: true, get: function () { return ensure_datatype_extracted_1.ensureDatatypeExtractedWithCallback; } });
__exportStar(require("../source/loader/generateAddressSpaceRaw"), exports);
var continuation_point_manager_1 = require("../source/continuation_points/continuation_point_manager");
Object.defineProperty(exports, "ContinuationPointManager", { enumerable: true, get: function () { return continuation_point_manager_1.ContinuationPointManager; } });
var finite_state_machine_1 = require("./state_machine/finite_state_machine");
Object.defineProperty(exports, "promoteToStateMachine", { enumerable: true, get: function () { return finite_state_machine_1.promoteToStateMachine; } });
__exportStar(require("../source/interfaces/state_machine/ua_transition_ex"), exports);
__exportStar(require("./namespace_impl"), exports);
__exportStar(require("./extension_object_array_node"), exports);
__exportStar(require("./event_data"), exports);
var nodeid_manager_1 = require("./nodeid_manager");
Object.defineProperty(exports, "NamespaceOptions", { enumerable: true, get: function () { return nodeid_manager_1.NamespaceOptions; } });
var nodeset_to_xml_1 = require("./nodeset_tools/nodeset_to_xml");
Object.defineProperty(exports, "dumpXml", { enumerable: true, get: function () { return nodeset_to_xml_1.dumpXml; } });
var typedictionary_to_xml_1 = require("./nodeset_tools/typedictionary_to_xml");
Object.defineProperty(exports, "dumpToBSD", { enumerable: true, get: function () { return typedictionary_to_xml_1.dumpToBSD; } });
var adjust_namespace_array_1 = require("./nodeset_tools/adjust_namespace_array");
Object.defineProperty(exports, "adjustNamespaceArray", { enumerable: true, get: function () { return adjust_namespace_array_1.adjustNamespaceArray; } });
var base_node_impl_1 = require("./base_node_impl");
Object.defineProperty(exports, "makeAttributeEventName", { enumerable: true, get: function () { return base_node_impl_1.makeAttributeEventName; } });
var reference_impl_1 = require("./reference_impl");
Object.defineProperty(exports, "resolveReferenceNode", { enumerable: true, get: function () { return reference_impl_1.resolveReferenceNode; } });
Object.defineProperty(exports, "resolveReferenceType", { enumerable: true, get: function () { return reference_impl_1.resolveReferenceType; } });
__exportStar(require("./data_access/check_variant_compatibility_ua_analog_item"), exports);
__exportStar(require("./data_access/add_dataItem_stuff"), exports);
__exportStar(require("./data_access/ua_multistate_discrete_impl"), exports);
__exportStar(require("./data_access/ua_multistate_value_discrete_impl"), exports);
__exportStar(require("./data_access/ua_two_state_discrete_impl"), exports);
var address_space_historical_data_node_1 = require("./historical_access/address_space_historical_data_node");
Object.defineProperty(exports, "VariableHistorian", { enumerable: true, get: function () { return address_space_historical_data_node_1.VariableHistorian; } });
var nodeid_manager_2 = require("../src/nodeid_manager");
Object.defineProperty(exports, "NodeIdManager", { enumerable: true, get: function () { return nodeid_manager_2.NodeIdManager; } });
__exportStar(require("./alarms_and_conditions"), exports);
__exportStar(require("./event_data"), exports);
__exportStar(require("node-opcua-nodeset-ua"), exports);
