// node-opcua-service-session/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignedSoftwareCertificate = exports.IssuedIdentityToken = exports.X509IdentityToken = exports.UserNameIdentityToken = exports.AnonymousIdentityToken = exports.CancelResponse = exports.CancelRequest = exports.CloseSessionResponse = exports.CloseSessionRequest = exports.ActivateSessionResponse = exports.ActivateSessionRequest = exports.CreateSessionResponse = exports.CreateSessionRequest = void 0;
/**
 * @module node-opcua-service-session
 */
var node_opcua_types_1 = require("node-opcua-types");
Object.defineProperty(exports, "CreateSessionRequest", { enumerable: true, get: function () { return node_opcua_types_1.CreateSessionRequest; } });
Object.defineProperty(exports, "CreateSessionResponse", { enumerable: true, get: function () { return node_opcua_types_1.CreateSessionResponse; } });
Object.defineProperty(exports, "ActivateSessionRequest", { enumerable: true, get: function () { return node_opcua_types_1.ActivateSessionRequest; } });
Object.defineProperty(exports, "ActivateSessionResponse", { enumerable: true, get: function () { return node_opcua_types_1.ActivateSessionResponse; } });
Object.defineProperty(exports, "CloseSessionRequest", { enumerable: true, get: function () { return node_opcua_types_1.CloseSessionRequest; } });
Object.defineProperty(exports, "CloseSessionResponse", { enumerable: true, get: function () { return node_opcua_types_1.CloseSessionResponse; } });
Object.defineProperty(exports, "CancelRequest", { enumerable: true, get: function () { return node_opcua_types_1.CancelRequest; } });
Object.defineProperty(exports, "CancelResponse", { enumerable: true, get: function () { return node_opcua_types_1.CancelResponse; } });
Object.defineProperty(exports, "AnonymousIdentityToken", { enumerable: true, get: function () { return node_opcua_types_1.AnonymousIdentityToken; } });
Object.defineProperty(exports, "UserNameIdentityToken", { enumerable: true, get: function () { return node_opcua_types_1.UserNameIdentityToken; } });
Object.defineProperty(exports, "X509IdentityToken", { enumerable: true, get: function () { return node_opcua_types_1.X509IdentityToken; } });
Object.defineProperty(exports, "IssuedIdentityToken", { enumerable: true, get: function () { return node_opcua_types_1.IssuedIdentityToken; } });
Object.defineProperty(exports, "SignedSoftwareCertificate", { enumerable: true, get: function () { return node_opcua_types_1.SignedSoftwareCertificate; } });
__exportStar(require("./SessionAuthenticationToken"), exports);
