// node-opcua-client/dist/client_utils.js
"use strict";
/**
 * @module node-opcua-client
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.readUAAnalogItem = void 0;
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_data_model_1 = require("node-opcua-data-model");
const node_opcua_nodeid_1 = require("node-opcua-nodeid");
const node_opcua_service_translate_browse_path_1 = require("node-opcua-service-translate-browse-path");
const node_opcua_status_code_1 = require("node-opcua-status-code");
const hasPropertyRefId = (0, node_opcua_nodeid_1.resolveNodeId)("HasProperty");
/* NodeId  ns=0;i=46*/
function browsePathPropertyRequest(nodeId, propertyName) {
    return new node_opcua_service_translate_browse_path_1.BrowsePath({
        relativePath: /* RelativePath   */ {
            elements: /* RelativePathElement */ [
                {
                    includeSubtypes: false,
                    isInverse: false,
                    referenceTypeId: hasPropertyRefId,
                    targetName: { namespaceIndex: 0, name: propertyName }
                }
            ]
        },
        startingNode: /* NodeId  */ nodeId
    });
}
function readUAAnalogItem(session, nodeId, ...args) {
    const callback = args[0];
    (0, node_opcua_assert_1.assert)(typeof callback === "function");
    const browsePath = [
        browsePathPropertyRequest(nodeId, "EngineeringUnits"),
        browsePathPropertyRequest(nodeId, "EURange"),
        browsePathPropertyRequest(nodeId, "InstrumentRange"),
        browsePathPropertyRequest(nodeId, "ValuePrecision"),
        browsePathPropertyRequest(nodeId, "Definition")
    ];
    const analogItemData = {
        definition: null,
        engineeringUnits: null,
        engineeringUnitsRange: null,
        instrumentRange: null,
        valuePrecision: null
    };
    session.translateBrowsePath(browsePath, (err, browsePathResults) => {
        if (err) {
            return callback(err);
        }
        browsePathResults = browsePathResults || [];
        const actions = [];
        const nodesToRead = [];
        function processProperty(browsePathResult, propertyName) {
            if (browsePathResult.statusCode === node_opcua_status_code_1.StatusCodes.Good) {
                browsePathResult.targets = browsePathResult.targets || [];
                nodesToRead.push({
                    attributeId: node_opcua_data_model_1.AttributeIds.Value,
                    nodeId: browsePathResult.targets[0].targetId
                });
                actions.push((readResult) => (analogItemData[propertyName] = readResult.value.value));
            }
        }
        processProperty(browsePathResults[0], "engineeringUnits");
        processProperty(browsePathResults[1], "engineeringUnitsRange");
        processProperty(browsePathResults[2], "instrumentRange");
        processProperty(browsePathResults[3], "valuePrecision");
        processProperty(browsePathResults[4], "definition");
        session.read(nodesToRead, (err1, dataValues) => {
            if (err1) {
                return callback(err1);
            }
            /* istanbul ignore next */
            if (!dataValues) {
                return callback(new Error("Internal Error"));
            }
            dataValues.forEach((result, index) => {
                actions[index].call(null, result);
            });
            callback(err1, analogItemData);
        });
    });
}
exports.readUAAnalogItem = readUAAnalogItem;
// tslint:disable:no-var-requires
// tslint:disable:max-line-length
const thenify = require("thenify");
const opts = { multiArgs: false };
module.exports.readUAAnalogItem = thenify.withCallback(module.exports.readUAAnalogItem, opts);
