// node-opcua-basic-types/dist/data_type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataType = void 0;
/***
 * @module node-opcua-basic-types
 */
var DataType;
(function (DataType) {
    DataType[DataType["Null"] = 0] = "Null";
    DataType[DataType["Boolean"] = 1] = "Boolean";
    DataType[DataType["SByte"] = 2] = "SByte";
    DataType[DataType["Byte"] = 3] = "Byte";
    DataType[DataType["Int16"] = 4] = "Int16";
    DataType[DataType["UInt16"] = 5] = "UInt16";
    DataType[DataType["Int32"] = 6] = "Int32";
    DataType[DataType["UInt32"] = 7] = "UInt32";
    DataType[DataType["Int64"] = 8] = "Int64";
    DataType[DataType["UInt64"] = 9] = "UInt64";
    DataType[DataType["Float"] = 10] = "Float";
    DataType[DataType["Double"] = 11] = "Double";
    DataType[DataType["String"] = 12] = "String";
    DataType[DataType["DateTime"] = 13] = "DateTime";
    DataType[DataType["Guid"] = 14] = "Guid";
    DataType[DataType["ByteString"] = 15] = "ByteString";
    DataType[DataType["XmlElement"] = 16] = "XmlElement";
    DataType[DataType["NodeId"] = 17] = "NodeId";
    DataType[DataType["ExpandedNodeId"] = 18] = "ExpandedNodeId";
    DataType[DataType["StatusCode"] = 19] = "StatusCode";
    DataType[DataType["QualifiedName"] = 20] = "QualifiedName";
    DataType[DataType["LocalizedText"] = 21] = "LocalizedText";
    DataType[DataType["ExtensionObject"] = 22] = "ExtensionObject";
    DataType[DataType["DataValue"] = 23] = "DataValue";
    DataType[DataType["Variant"] = 24] = "Variant";
    DataType[DataType["DiagnosticInfo"] = 25] = "DiagnosticInfo";
})(DataType = exports.DataType || (exports.DataType = {}));
