// node-opcua-nodeset-ua/dist/enum_redundancy_support.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumRedundancySupport = void 0;
/**
 * |           |                                                  |
 * |-----------|--------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                      |
 * | nodeClass |DataType                                          |
 * | name      |RedundancySupport                                 |
 * | isAbstract|false                                             |
 */
var EnumRedundancySupport;
(function (EnumRedundancySupport) {
    EnumRedundancySupport[EnumRedundancySupport["None"] = 0] = "None";
    EnumRedundancySupport[EnumRedundancySupport["Cold"] = 1] = "Cold";
    EnumRedundancySupport[EnumRedundancySupport["Warm"] = 2] = "Warm";
    EnumRedundancySupport[EnumRedundancySupport["Hot"] = 3] = "Hot";
    EnumRedundancySupport[EnumRedundancySupport["Transparent"] = 4] = "Transparent";
    EnumRedundancySupport[EnumRedundancySupport["HotAndMirrored"] = 5] = "HotAndMirrored";
})(EnumRedundancySupport = exports.EnumRedundancySupport || (exports.EnumRedundancySupport = {}));
