// node-opcua-nodeset-ua/dist/enum_filter_operator.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumFilterOperator = void 0;
/**
 * |           |                                                  |
 * |-----------|--------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                      |
 * | nodeClass |DataType                                          |
 * | name      |FilterOperator                                    |
 * | isAbstract|false                                             |
 */
var EnumFilterOperator;
(function (EnumFilterOperator) {
    EnumFilterOperator[EnumFilterOperator["Equals"] = 0] = "Equals";
    EnumFilterOperator[EnumFilterOperator["IsNull"] = 1] = "IsNull";
    EnumFilterOperator[EnumFilterOperator["GreaterThan"] = 2] = "GreaterThan";
    EnumFilterOperator[EnumFilterOperator["LessThan"] = 3] = "LessThan";
    EnumFilterOperator[EnumFilterOperator["GreaterThanOrEqual"] = 4] = "GreaterThanOrEqual";
    EnumFilterOperator[EnumFilterOperator["LessThanOrEqual"] = 5] = "LessThanOrEqual";
    EnumFilterOperator[EnumFilterOperator["Like"] = 6] = "Like";
    EnumFilterOperator[EnumFilterOperator["Not"] = 7] = "Not";
    EnumFilterOperator[EnumFilterOperator["Between"] = 8] = "Between";
    EnumFilterOperator[EnumFilterOperator["InList"] = 9] = "InList";
    EnumFilterOperator[EnumFilterOperator["And"] = 10] = "And";
    EnumFilterOperator[EnumFilterOperator["Or"] = 11] = "Or";
    EnumFilterOperator[EnumFilterOperator["Cast"] = 12] = "Cast";
    EnumFilterOperator[EnumFilterOperator["InView"] = 13] = "InView";
    EnumFilterOperator[EnumFilterOperator["OfType"] = 14] = "OfType";
    EnumFilterOperator[EnumFilterOperator["RelatedTo"] = 15] = "RelatedTo";
    EnumFilterOperator[EnumFilterOperator["BitwiseAnd"] = 16] = "BitwiseAnd";
    EnumFilterOperator[EnumFilterOperator["BitwiseOr"] = 17] = "BitwiseOr";
})(EnumFilterOperator = exports.EnumFilterOperator || (exports.EnumFilterOperator = {}));
