// node-opcua-nodeset-ua/dist/enum_trust_list_masks.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumTrustListMasks = void 0;
/**
 * |           |                                                  |
 * |-----------|--------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                      |
 * | nodeClass |DataType                                          |
 * | name      |TrustListMasks                                    |
 * | isAbstract|false                                             |
 */
var EnumTrustListMasks;
(function (EnumTrustListMasks) {
    EnumTrustListMasks[EnumTrustListMasks["None"] = 0] = "None";
    EnumTrustListMasks[EnumTrustListMasks["TrustedCertificates"] = 1] = "TrustedCertificates";
    EnumTrustListMasks[EnumTrustListMasks["TrustedCrls"] = 2] = "TrustedCrls";
    EnumTrustListMasks[EnumTrustListMasks["IssuerCertificates"] = 4] = "IssuerCertificates";
    EnumTrustListMasks[EnumTrustListMasks["IssuerCrls"] = 8] = "IssuerCrls";
    EnumTrustListMasks[EnumTrustListMasks["All"] = 15] = "All";
})(EnumTrustListMasks = exports.EnumTrustListMasks || (exports.EnumTrustListMasks = {}));
