// node-opcua-nodeset-ua/dist/enum_node_class.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumNodeClass = void 0;
/**
 * |           |                                                  |
 * |-----------|--------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                      |
 * | nodeClass |DataType                                          |
 * | name      |NodeClass                                         |
 * | isAbstract|false                                             |
 */
var EnumNodeClass;
(function (EnumNodeClass) {
    /**
     * No value is specified.
     */
    EnumNodeClass[EnumNodeClass["Unspecified"] = 0] = "Unspecified";
    /**
     * The Node is an Object.
     */
    EnumNodeClass[EnumNodeClass["Object"] = 1] = "Object";
    /**
     * The Node is a Variable.
     */
    EnumNodeClass[EnumNodeClass["Variable"] = 2] = "Variable";
    /**
     * The Node is a Method.
     */
    EnumNodeClass[EnumNodeClass["Method"] = 4] = "Method";
    /**
     * The Node is an ObjectType.
     */
    EnumNodeClass[EnumNodeClass["ObjectType"] = 8] = "ObjectType";
    /**
     * The Node is a VariableType.
     */
    EnumNodeClass[EnumNodeClass["VariableType"] = 16] = "VariableType";
    /**
     * The Node is a ReferenceType.
     */
    EnumNodeClass[EnumNodeClass["ReferenceType"] = 32] = "ReferenceType";
    /**
     * The Node is a DataType.
     */
    EnumNodeClass[EnumNodeClass["DataType"] = 64] = "DataType";
    /**
     * The Node is a View.
     */
    EnumNodeClass[EnumNodeClass["View"] = 128] = "View";
})(EnumNodeClass = exports.EnumNodeClass || (exports.EnumNodeClass = {}));
